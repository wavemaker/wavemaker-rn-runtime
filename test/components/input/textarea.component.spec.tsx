import React, { createRef } from 'react';
import {
  act,
  cleanup,
  fireEvent,
  render,
  waitFor,
} from '@testing-library/react-native';
import WmTextarea from '@wavemaker/app-rn-runtime/components/input/textarea/textarea.component';
import { Platform, TextInput } from 'react-native';

const defaultProps = {
  id: 'wmTextarea',
  name: 'wmTextarea',
  placeholder: 'Place your text',
  disabled: false,
  show: true,
  styles: null,
  classname: null,
  listener: null,
  showindevice: null,
  showskeleton: false,
  deferload: false,
};

describe('Test Textarea component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    jest.useFakeTimers();
  });

  afterEach(() => {
    // jest.runOnlyPendingTimers();
    jest.useRealTimers();
    cleanup();
  });

  test('should render correctly with default props', () => {
    const tree = render(<WmTextarea {...defaultProps} />);
    expect(tree.getByPlaceholderText('Place your text')).toBeTruthy();
    expect(tree).toMatchSnapshot();
  });

  test('should show floating label when focused', () => {
    const tree = render(
      <WmTextarea {...defaultProps} floatinglabel="sample label" />
    );
    const { queryByText, UNSAFE_getByType } = tree;
    const input = UNSAFE_getByType(TextInput);

    fireEvent(input, 'focus', {
      target: null,
    });

    expect(queryByText('sample label')).toBeTruthy();
  });

  test('should call onChangeText with correct value', () => {
    const onChangeTextMock = jest.spyOn(WmTextarea.prototype, 'onChangeText');
    const { getByPlaceholderText } = render(<WmTextarea {...defaultProps} />);
    const input = getByPlaceholderText('Place your text');
    const text = 'new text';

    fireEvent.changeText(input, text);

    expect(onChangeTextMock).toHaveBeenCalledWith(text);
  });

  test('should handle disabled prop correctly', () => {
    const { getByPlaceholderText } = render(
      <WmTextarea {...defaultProps} disabled={true} />
    );
    const input = getByPlaceholderText('Place your text');

    expect(input.props.editable).toBe(false);
  });

  test('should handle readonly prop correctly', () => {
    const { getByPlaceholderText } = render(
      <WmTextarea {...defaultProps} readonly={true} />
    );
    const input = getByPlaceholderText('Place your text');

    expect(input.props.editable).toBe(false);
  });

  test('should handle maxchars prop correctly', () => {
    const maxchars = 100;
    const { getByPlaceholderText } = render(
      <WmTextarea {...defaultProps} maxchars={maxchars} />
    );
    const input = getByPlaceholderText('Place your text');

    expect(input.props.maxLength).toBe(maxchars);
  });

  test('should have default value for native platform', () => {
    const tree = render(
      <WmTextarea {...defaultProps} datavalue="sample text" />
    );
    expect(tree.UNSAFE_getByType(TextInput).props.defaultValue).toBe(
      'sample text'
    );
    expect(tree).toMatchSnapshot();
  });

  test('should have default value for web platform', () => {
    (Platform as any).OS = 'web';

    const tree = render(
      <WmTextarea {...defaultProps} datavalue="sample text" />
    );
    expect(tree.UNSAFE_getByType(TextInput).props.value).toBe('sample text');
    expect(tree).toMatchSnapshot();
  });

  test('should render field in invalid state when isValid is false', async () => {
    const invalidStyle = { borderColor: 'red' };
    const { getByPlaceholderText } = render(
      <WmTextarea {...defaultProps} styles={{ invalid: invalidStyle }} />
    );
    const input = getByPlaceholderText('Place your text');
    fireEvent.changeText(input, 'text exceeding validation');

    await waitFor(() => {
      const styleArr = input.props.style[0];
      const style = {};
      styleArr.forEach((item) => {
        if (!item) return;
        Object.keys(item).forEach((key) => {
          style[key] = item[key];
        });
      });

      expect(style).toMatchObject(invalidStyle);
    });
  });

  test('should render floating label with labelformat', () => {
    const { getByText, UNSAFE_getByType } = render(
      <WmTextarea {...defaultProps} floatinglabel="Test Label" />
    );
    const input = UNSAFE_getByType(TextInput);
    expect(getByText('Test Label')).toBeTruthy();
  });

  test('should hide skeleton when showSkeleton is false', () => {
    const { queryByTestId } = render(
      <WmTextarea {...defaultProps} showskeleton={false} />
    );
    const skeleton = queryByTestId('skeleton');
    expect(skeleton).toBeNull();
  });

  test('renders with accessibility properties', async () => {
    const { getByLabelText } = render(
      <WmTextarea {...defaultProps} accessibilitylabel="Textarea input" />
    );

    await waitFor(() => {
      const input = getByLabelText('Textarea input');
      expect(input.props.accessibilityLabel).toBe('Textarea input');
    });
  });

  test('should auto complete username if autocomplete prop is provided', () => {
    const { getByPlaceholderText } = render(
      <WmTextarea {...defaultProps} autocomplete={true} />
    );
    const input = getByPlaceholderText('Place your text');
    expect(input.props.autoComplete).toBe('username');
  });

  test('should auto complete off if autocomplete prop is not provided or falsy', () => {
    const { getByPlaceholderText } = render(
      <WmTextarea {...defaultProps} autocomplete={false} />
    );
    const input = getByPlaceholderText('Place your text');
    expect(input.props.autoComplete).toBe('off');
  });

  test('should have keyboardType as numeric for type as number', () => {
    const { getByPlaceholderText } = render(
      <WmTextarea {...defaultProps} type="number" />
    );

    const input = getByPlaceholderText('Place your text');
    expect(input.props.keyboardType).toBe('numeric');
  });

  test('should have keyboardType as phone-pad for type as tel', () => {
    const { getByPlaceholderText } = render(
      <WmTextarea {...defaultProps} type="tel" />
    );

    const input = getByPlaceholderText('Place your text');
    expect(input.props.keyboardType).toBe('phone-pad');
  });

  test('should have keyboardType as email-address for type as email', () => {
    const { getByPlaceholderText } = render(
      <WmTextarea {...defaultProps} type="email" />
    );

    const input = getByPlaceholderText('Place your text');
    expect(input.props.keyboardType).toBe('email-address');
  });

  test('should not show component when show prop is false', () => {
    const tree = render(<WmTextarea {...defaultProps} show={false} />);

    const styleArr =
      tree.getByPlaceholderText('Place your text').props.style[0];
    const style = {};
    styleArr.forEach((item) => {
      if (!item) return;
      Object.keys(item).forEach((key) => {
        style[key] = item[key];
      });
    });

    expect(style).toMatchObject({
      height: 0,
      width: 0,
    });
  });

  test('should render with default placeholder when placeholder passed as props is falsy', () => {
    const tree = render(<WmTextarea {...defaultProps} placeholder="" />);

    expect(tree.getByPlaceholderText('Place your text')).toBeTruthy();
  });
});
