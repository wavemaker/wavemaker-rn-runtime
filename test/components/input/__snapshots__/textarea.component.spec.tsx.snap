// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Textarea component should have default value for native platform 1`] = `
<TextInput
  accessibilityState={
    {
      "disabled": false,
    }
  }
  accessible={true}
  activeFloatingLabelStyle={{}}
  allowContentSelection={true}
  autoComplete="username"
  autoFocus={null}
  background={
    <React.Fragment>
      <BackgroundComponent
        style={
          {
            "borderRadius": 6,
          }
        }
      />
    </React.Fragment>
  }
  caretHidden={false}
  contextMenuHidden={false}
  defaultValue="sample text"
  editable={true}
  floatingLabelStyle={{}}
  isInputFocused={false}
  maxLength={null}
  multiline={true}
  numberOfLines={4}
  onBlur={[Function]}
  onChange={[Function]}
  onChangeText={[Function]}
  onFocus={[Function]}
  onKeyPress={[Function]}
  onSelectionChange={[Function]}
  placeholder="Place your text"
  placeholderTextColor="#bbbbbb"
  selection={null}
  style={
    [
      [
        {
          "backgroundColor": "#ffffff",
          "borderColor": "#dddddd",
          "borderRadius": 6,
          "borderStyle": "solid",
          "borderWidth": 1,
          "fontFamily": "Roboto",
          "minHeight": 160,
          "paddingBottom": 12,
          "paddingLeft": 12,
          "paddingRight": 12,
          "paddingTop": 12,
        },
        {
          "fontFamily": "Roboto",
          "fontSize": 16,
          "textAlignVertical": "top",
          "userSelect": "text",
        },
        {},
        {},
      ],
      {},
    ]
  }
  testID="wmTextarea_i"
/>
`;

exports[`Test Textarea component should have default value for web platform 1`] = `
<TextInput
  accessibilityLabel="wmTextarea_i"
  activeFloatingLabelStyle={{}}
  allowContentSelection={true}
  autoComplete="username"
  autoFocus={null}
  background={
    <React.Fragment>
      <BackgroundComponent
        style={
          {
            "borderRadius": 6,
          }
        }
      />
    </React.Fragment>
  }
  caretHidden={false}
  contextMenuHidden={false}
  editable={true}
  floatingLabelStyle={{}}
  isInputFocused={false}
  maxLength={null}
  multiline={true}
  numberOfLines={4}
  onBlur={[Function]}
  onChange={[Function]}
  onChangeText={[Function]}
  onFocus={[Function]}
  onKeyPress={[Function]}
  onSelectionChange={[Function]}
  placeholder="Place your text"
  placeholderTextColor="#bbbbbb"
  selection={null}
  style={
    [
      [
        {
          "backgroundColor": "#ffffff",
          "borderColor": "#dddddd",
          "borderRadius": 6,
          "borderStyle": "solid",
          "borderWidth": 1,
          "fontFamily": "Roboto",
          "minHeight": 160,
          "paddingBottom": 12,
          "paddingLeft": 12,
          "paddingRight": 12,
          "paddingTop": 12,
        },
        {
          "fontFamily": "Roboto",
          "fontSize": 16,
          "textAlignVertical": "top",
          "userSelect": "text",
        },
        {},
        {},
      ],
      {},
    ]
  }
  testID="wmTextarea_i"
  value="sample text"
/>
`;

exports[`Test Textarea component should render correctly with default props 1`] = `
<TextInput
  accessibilityState={
    {
      "disabled": false,
    }
  }
  accessible={true}
  activeFloatingLabelStyle={{}}
  allowContentSelection={true}
  autoComplete="username"
  autoFocus={null}
  background={
    <React.Fragment>
      <BackgroundComponent
        style={
          {
            "borderRadius": 6,
          }
        }
      />
    </React.Fragment>
  }
  caretHidden={false}
  contextMenuHidden={false}
  defaultValue=""
  editable={true}
  floatingLabelStyle={{}}
  isInputFocused={false}
  maxLength={null}
  multiline={true}
  numberOfLines={4}
  onBlur={[Function]}
  onChange={[Function]}
  onChangeText={[Function]}
  onFocus={[Function]}
  onKeyPress={[Function]}
  onSelectionChange={[Function]}
  placeholder="Place your text"
  placeholderTextColor="#bbbbbb"
  selection={null}
  style={
    [
      [
        {
          "backgroundColor": "#ffffff",
          "borderColor": "#dddddd",
          "borderRadius": 6,
          "borderStyle": "solid",
          "borderWidth": 1,
          "fontFamily": "Roboto",
          "minHeight": 160,
          "paddingBottom": 12,
          "paddingLeft": 12,
          "paddingRight": 12,
          "paddingTop": 12,
        },
        {
          "fontFamily": "Roboto",
          "fontSize": 16,
          "textAlignVertical": "top",
          "userSelect": "text",
        },
        {},
        {},
      ],
      {},
    ]
  }
  testID="wmTextarea_i"
/>
`;
